### PR TITLE
http json-rpc

### PR DESCRIPTION
**Pull request by xrdavies**
http json-rpc, supports json-rpc specification 1.0, 1.1 and 2.0.

_**Only accept http request from local node. Trusted white list will be in next version.**_

Available methods:
**xdag_do_xfer** Do transfer.
**xdag_get_account** Get information of N accounts. Default is 20.
**xdag_get_balance** Get balance of specific account, if no account specified, it show total balance.
**xdag_state** Get current state of xdag.
**xdag_stats** Get current hash rate of xdag.

**xdag_do_xfer**
request:
```
{
	"method" : "xdag_do_xfer",
	"params" : [{
		"amount" : "0.1",
		"address" : "FQglVQtb60vQv2DOWEUL7yh3smtj7g1s"
	}],
	"id" : 1
}
```
response:
```
{
	"result" : [{
		"block" : "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
	}],
	"error" : null, 
	"id" : 1
}

```
**xdag_get_account**
request:
```
{
	"method" : "xdag_get_account",
	"params" : ["20"],
	"id" : 1
}

```
response:
```
{
	"result" : [
		{
			"address" : "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
			"balance" : "100000.00001",
			"key" : "0"
		},
		{
			"address" : "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
			"balance" : "200000.00001",
			"key" : "0"
		}],
	"error" : null, 
	"id" : 1
}
```

**xdag_get_balance**
request:
```
{
	"method" : "xdag_get_balance",
	"params" : ["AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"],
	"id" : 1
}
```

response:
```
{
	"result" : [{
			"balance" : "100000.00001"
		}],
	"error" : null, 
	"id" : 1
}
```

**xdag_state**
request:
```
{
	"method" : "xdag_state",
	"params" : [],
	"id" : 1
}
```

response:
```
{
	"result" : ["current state"],
	"error" : null, 
	"id" : 1
}
```

**xdag_stats**
request:
```
{
	"method" : "xdag_stats",
	"params" : [],
	"id" : 1
}
```

response:
```
{
	"result" : ["current hashrate"],
	"error" : null, 
	"id" : 1
}
```

### CODE REVIEW

File: client/json-rpc/rpc_procedures.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		
+		if(strlen(address)) {
+			xdag_address2hash(address, hash);
+			balance = xdag_get_balance(hash);
+		} else {
+			balance = xdag_get_balance(0);
+		}
+		
+		char str[128] = {0};
+		sprintf(str, "%.9Lf",  xdag_amount2xdag(balance) + (long double)xdag_amount2cheato(balance) / 1000000000);
```
Comment (jonano614) [inline client/json-rpc/rpc_procedures.c]: You moved declaration of amount2xdags to the header file. You could use existing function 
long double amount2xdags(xdag_amount_t amount)
{
	return xdag_amount2xdag(amount) + (long double)xdag_amount2cheato(amount) / 1000000000;
}

File: client/json-rpc/rpc_procedures.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+			ctx->error_message = strdup("Invalid parameters.");
+			return NULL;
+		}
+	} else {
+		ctx->error_code = 1;
+		ctx->error_message = strdup("Invalid parameters.");
+		return NULL;
+	}
+	
+	if(g_xdag_state < XDAG_STATE_XFER) {
```
Comment (jonano614) [inline client/json-rpc/rpc_procedures.c]: I really feel like we need OOP :) Most of this logic is similar to command.c. Something like base - derived classes.

File: client/json-rpc/rpc_procedures.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+	}
+}
+
+#define rpc_register_func(command) xdag_rpc_service_register_procedure(&method_##command, #command, NULL);
+
+/* init rpc procedures */
+int xdag_rpc_init_procedures(void)
+{
+	/* register query func */
+	rpc_register_func(account);
```
Comment (jonano614) [inline client/json-rpc/rpc_procedures.c]: How these functions work?

File: client/init.c [hunk 4]
```diff
@@ -67,6 +68,9 @@ int xdag_init(int argc, char **argv, int isGui)
 	
 	/* initialize log system */
 	if (xdag_log_init()) return -1;
+	
+	/* initialize json rpc */
+	if (xdag_rpc_service_init()) return -1;
```
[the remark this answers is not in the document]
Comment (xrdavies) [reply client/init.c]: Yes, it's in plan scheduled after long parameter feature.  I would prefer to re-factory the parameter parsing. It's quite inconvenient to add long parameters.

File: client/json-rpc/rpc_procedures.c (hunk 1 above)
Comment (xrdavies) [reply client/json-rpc/rpc_procedures.c]: Good point!

File: client/json-rpc/rpc_procedures.c (hunk 3 above)
Comment (xrdavies) [reply client/json-rpc/rpc_procedures.c]: @jonano614 

Removed useless procedures, if we need it we can add them later.
Only keep five procedures.
They are described in PR description.

File: client/json-rpc/rpc_procedures.c (hunk 1 above)
Comment (jonano614) [reply client/json-rpc/rpc_procedures.c]: Will you fix this line?

File: client/json-rpc/rpc_procedures.c (hunk 2 above)
Comment (sgaragagghu) [reply client/json-rpc/rpc_procedures.c]: You can do OOP on C too :D

**Pull request merged**